Wdb 89/create reset password page

kibartas opened the pull request:
This PR has everything you'd ever want from a PR.

- A change to an endpoint that some people absolutely despise
- Bugs (probably)
- Reset password page and card
- Other changes, like:
1. API path change 
2. ChangeProfileForm now allows changing only email or password

Commit:
[WDB-89]
change UsersControllers API endpoints forgot-password and update-credentials to be more robust

Commit:
[WDB-89]
straighten out the API paths and their parameters in front

Commit:
[WDB-89]
ChangeProfileForm now lets user only update their email

Commit:
[WDB-89]
minor grid improvements in ForgotPasswordCard

Commit:
[WDB-89]
create and finish ResetPasswordCard + page

Commit:
[WDB-89]
ProfilePage now displays if the user inputs a wrong password

Commit:
[WDB-89]
revert accidental appsettings.json changes

Commit:
[WDB-89]
merge master

salmonela7 commented on backend/backend/Controllers/UsersController.cs:
```diff
@@ -191,36 +191,74 @@ public ActionResult ResetPassword(string newPassword)
+
+            if (!string.IsNullOrWhiteSpace(request.Email)) user.Email = request.Email;
+            
+            if (string.IsNullOrWhiteSpace(request.OldPassword) ^ string.IsNullOrWhiteSpace(request.NewPassword))
+                return BadRequest("Only one password field provided");
```
Delete this since we check it in line 232-236 :D

Commit:
[WDB-89]
remove redundant check from users/id endpoint (bye bye XOR ;( )

tmaciulis22 committed:
[WDB-89] Layout changes

tmaciulis22 commented on backend/backend/Controllers/UsersController.cs:
```diff
@@ -191,36 +191,71 @@ public ActionResult ResetPassword(string newPassword)
-            if (!Guid.TryParse(userIdClaim.Value, out var userId)) return NotFound();
+            if (userIdClaim is null) return NotFound("JWT doesn't correspond to any user");
+            if (!Guid.TryParse(userIdClaim.Value, out var userId)) return NotFound("JWT token is malformed");
+            if (id != userId) return BadRequest("Something wrong has happened");
```
Maybe we should write here, that user id's were mismatched or something similar

tmaciulis22 commented on backend/backend/Controllers/UsersController.cs:
```diff
@@ -191,36 +191,71 @@ public ActionResult ResetPassword(string newPassword)
+
+            if (!string.IsNullOrWhiteSpace(request.Email)) user.Email = request.Email;
+            
+            if ((string.IsNullOrWhiteSpace(request.OldPassword) && !string.IsNullOrWhiteSpace(request.NewPassword)) 
+                || !string.IsNullOrWhiteSpace(request.OldPassword) && string.IsNullOrWhiteSpace(request.NewPassword))
```
```suggestion
            if (string.IsNullOrWhiteSpace(request.OldPassword) || string.IsNullOrWhiteSpace(request.NewPassword))
```
Wouldn't this do the same?

kibartas replied:
This here is exactly about the case where one of the fields is empty and the other one ain't. That means that the request is malformed and we should throw it out

In your case, there is a possibility that both of them are null, and in that case it doesn't necessarily mean that the request is bad

tmaciulis22 replied:
But haven't we already covered email case above? Then my suggestion would be enough to cover the password cases. I mean to update password we need both of them: old and new, so wouldn't this suggested check be enough? I might be confused here, maybe you are right ✌🏻

tmaciulis22 replied:
Wait since this endpoint is used for both Aurimas' profile page and yours Reset, right? So that means we need to only check either if both old and new is provided, or only new?

kibartas replied:
Yes, I thought about that, but I'm saying that if one of them IS provided, then the request is malformed and we should discard it altogether, without updating email or w/e

No, different endpoints are used. It's kinda weird, I know, but reset password needs JWT Reset token, while this one doesn't need it.

tmaciulis22 commented on backend/backend/Controllers/UsersController.cs:
```diff
@@ -191,36 +191,71 @@ public ActionResult ResetPassword(string newPassword)
+            if (!string.IsNullOrWhiteSpace(request.OldPassword) && !string.IsNullOrWhiteSpace(request.NewPassword))
+            {
+                if (request.OldPassword == request.NewPassword) 
+                    return BadRequest("Old password is the same as the new");
```
```suggestion
                    return BadRequest("New password is the same as the old");
```
Just a very nitpicky suggestion, but this seems more natural in my opinion

tmaciulis22 replied:
Would understand if you will say that it's a stupid suggestion ✌🏻

kibartas replied:
I really don't care, if it makes you happy

tmaciulis22 commented on backend/backend/Controllers/UsersController.cs:
```diff
@@ -191,36 +191,71 @@ public ActionResult ResetPassword(string newPassword)
 
-            user.Email = request.Email;
-            user.SetNewPassword(request.NewPassword);
+            if (string.IsNullOrWhiteSpace(request.Email) && string.IsNullOrWhiteSpace(request.OldPassword)
+                                                         && string.IsNullOrWhiteSpace(request.NewPassword))
+            {
+                return BadRequest("Request is empty");
+            }
```
Shouldn't this check be on top of the method? This way we would save unnecessary processing for backend with all those claims etc

kibartas replied:
True that

tmaciulis22 commented on frontend/src/api/PublicAPI.js:
```diff
@@ -24,3 +24,9 @@ export const sendForgotPasswordEmail = (email) =>
 export const getCurrentUser = (token) =>
   Api.get('/users/current', { headers: { Authorization: `Bearer ${token}` } });
+
+export const resetPassword = (token, newPassword) =>
```
I might be too tired to follow logic here, but don't we need to pass `oldPassword` as well?

kibartas replied:
Reset password only requires the user's Reset JWT token and their new password, because it is used after the "Forgot password?"

Maybe you're confusing this one with Aurimas' updateCredentials which he did yesterday. There we require the user to enter the old password

tmaciulis22 replied:
Wait so we aren't using the same endpoint for update credentials and reset password?

kibartas replied:
No

tmaciulis22 commented on frontend/src/components/ChangeProfileForm/ChangeProfileForm.js:
```diff
@@ -50,17 +50,19 @@ export default function ChangeProfileForm({
-    if (newPassword === password) {
-      setShowError({ ...showError, errorNewPasswordMatch: true });
-      return;
+    if (password !== '' && newPassword !== '') {
```
Is this necessary?

kibartas replied:
If both of them are not empty then we check if they are regex valid. If they are both empty, we ignore the regexes

tmaciulis22 replied:
But doesn't regex do the same thing?

kibartas replied:
No, I don't think so

tmaciulis22 commented on frontend/src/components/ResetPasswordCard/ResetPasswordCard.js:
```diff
@@ -0,0 +1,187 @@
+} from '../../assets';
+
+const ResetPasswordCard = ({ onSend }) => {
+  const [showPasswordError, setShowPasswordError] = useState(false);
```
Move this state to bottom of states. Move `showConfirmPasswordError` as well. This way we will keep the logical grouping of states. First come the values, then errors, etc.
I know it's another nitpicky suggestion. Feel free to object.

kibartas replied:
It's fine, it's a quick change

tmaciulis22 reviewed (changes requested):
🔍 
Also, two suggestions:
1. I think we should add timeout to forgot password form requests, so the user won't spam our email service.
2. I think we should redirect the user to login after successful password reset. Problem might be with success snackbar, so maybe `onClose` of snackbar a redirection should happen.

kibartas committed:
Update backend/backend/Controllers/UsersController.cs

Co-authored-by: Tautvydas Mačiulis <maciulis.tautvydas22@gmail.com>

Commit:
[WDB-89]
change the ordering of useState in ResetPasswordCard

Commit:
[WDB-89]
move if statement in UsersController

Commit:
[WDB-89]
more expressive error message in UsersController

Commit:
Merge branch 'WDB-89/Create-reset-password-page' of github.com:kibartas/PSK-Front-End into WDB-89/Create-reset-password-page

kibartas commented:
> mag
> Also, two suggestions:
> 
>     1. I think we should add timeout to forgot password form requests, so the user won't spam our email service.
> 
>     2. I think we should redirect the user to login after successful password reset. Problem might be with success snackbar, so maybe `onClose` of snackbar a redirection should happen.

1. I'm pretty sure that's out of the scope of our task. If you're talking about rapid requests, then that would probably be the reverse proxy's concern. If you're talking about the user just clicking the same button a bunch of times for giggles, then I propose disabling the button after request has been sent. We could do, idk, an IP timeout if we wanted to, for such things, but that seems, you know, out of scope

2. Small thing. Sure

Commit:
[WDB-89]
sneaky user.Confirmed check added in reset password

Commit:
[WDB-89]
add timeout and redirect after resetting password

Commit:
[WDB-89]
disable Send button in forget password when the request is being sent and on success

Commit:
[WDB-89]
make some buttons disabled

tmaciulis22 reviewed (approved):
👏🏻👏🏻👏🏻
